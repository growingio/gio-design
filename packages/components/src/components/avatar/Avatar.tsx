import React, { useState, useEffect, useRef } from 'react';
import classNames from 'classnames';
import { More } from '@gio-design/icons';
import { isNil } from 'lodash';
import Tooltip from '../tooltip';
import { AvatarProps } from './interfaces';
import usePrefixCls from '../../utils/hooks/use-prefix-cls';
import composeRef from '../../utils/composeRef';

const Avatar = React.forwardRef<HTMLSpanElement, AvatarProps>((props: AvatarProps, ref: React.Ref<HTMLSpanElement>) => {
  const {
    className,
    size = 'default',
    droppable = false,
    children: userName,
    src,
    omit = true,
    displayTooltip = false,
    prefixCls: customizePrefixCls,
    placement = 'bottom',
    tooltipTitle,
    ...rest
  } = props;

  const [isImgExist, setIsImgExist] = useState<boolean>(src !== undefined);
  const [scale, setScale] = useState<number>(1);
  const nodeRef = useRef<HTMLSpanElement>(null);
  const childrenRef = useRef<HTMLSpanElement>(null);
  const mergedRef = composeRef(ref, nodeRef);

  useEffect(() => {
    if (nodeRef.current && childrenRef.current) {
      const nodeWidth = nodeRef.current.offsetWidth;
      const childrenWidth = childrenRef.current.offsetWidth;
      setScale(nodeWidth - 8 < childrenWidth ? (nodeWidth - 8) / childrenWidth : 1);
    }
  }, [userName]);

  const prefixCls = usePrefixCls('avatar', customizePrefixCls);
  const classString = classNames(className, prefixCls, {
    [`${prefixCls}-sm`]: size === 'small',
    [`${prefixCls}-df`]: size === 'default',
    [`${prefixCls}-lg`]: size === 'large',
    [`${prefixCls}-hg`]: size === 'huge',
  });

  const childrenStyle: React.CSSProperties = {
    transform: `scale(${scale})`,
  };

  const renderMore = () => {
    if (droppable) {
      return (
        <div className={classNames({ [`${prefixCls}-droppable`]: droppable })}>
          <More size="16px" color="#FFFFFF" />
        </div>
      );
    }
    return null;
  };

  const renderAvatar = () => {
    if (!!src && isImgExist) {
      return <img alt="avatar" src={src} onError={() => setIsImgExist(false)} />;
    }
    if (userName !== undefined && typeof userName === 'string') {
      const prefixUserName = omit && typeof userName === 'string' ? userName.trim()[0] : userName.trim();
      return (
        <span ref={childrenRef} style={childrenStyle}>
          {prefixUserName}
        </span>
      );
    }
    return null;
  };

  const renderTooltip = (child: React.ReactElement) =>
    displayTooltip ? (
      <Tooltip title={isNil(tooltipTitle) ? userName?.trim() : tooltipTitle} placement={placement}>
        {child}
      </Tooltip>
    ) : (
      child
    );

  return renderTooltip(
    // For Dropdown trigger will set Event on rest
    // eslint-disable-next-line react/jsx-props-no-spreading
    <span ref={mergedRef} className={classString} {...rest}>
      {renderMore()}
      {renderAvatar()}
    </span>
  );
});

export default Avatar;
