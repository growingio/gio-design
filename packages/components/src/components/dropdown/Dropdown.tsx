import React, { useContext } from 'react';
import Tooltip from '../tooltip';
import { DropdownProps } from './interface';
import { ConfigContext } from '../config-provider';
import useControlledState from '../../utils/hooks/useControlledState';

const Dropdown = (props: DropdownProps) => {
  const {
    children,
    prefixCls: customizePrefixCls,
    placement = 'bottom',
    trigger = 'click',
    visible,
    onVisibleChange,
    ...rest
  } = props;
  const { getPrefixCls } = useContext(ConfigContext);
  const prefixCls = getPrefixCls('dropdown', customizePrefixCls);
  const [controlledVisible, setControlledVisible] = useControlledState(visible, false);

  return (
    <Tooltip
      prefixCls={prefixCls}
      trigger={trigger}
      placement={placement}
      visible={controlledVisible}
      onVisibleChange={(_visible) => {
        setControlledVisible(_visible);
        onVisibleChange?.(_visible);
      }}
      {...rest}
    >
      {children}
    </Tooltip>
  );
};

export default Dropdown;
