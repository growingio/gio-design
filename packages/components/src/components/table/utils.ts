import { clone, has, get, join, set } from 'lodash';
import { ColumnsType, InnerColumnsType } from './interface';

export const translateInnerColumns = <RecordType>(columns: ColumnsType<RecordType>): InnerColumnsType<RecordType> =>
  clone(columns).map((cloneColumn) => {
    if (!has(cloneColumn, 'key')) {
      if (has(cloneColumn, 'dataIndex')) {
        if (Array.isArray(get(cloneColumn, 'dataIndex'))) {
          cloneColumn.key = join(get(cloneColumn, 'dataIndex'), '-');
        } else {
          cloneColumn.key = get(cloneColumn, 'dataIndex');
        }
      } else {
        console.warn('gio-design table: column key or dataIndex must have one');
      }
    }
    if (has(cloneColumn, 'children')) {
      set(cloneColumn, 'children', translateInnerColumns(get(cloneColumn, 'children')));
    }
    return cloneColumn;
  }) as InnerColumnsType<RecordType>;

export default {}
