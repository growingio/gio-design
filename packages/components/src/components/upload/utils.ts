import { defaultRootPrefixCls } from '../config-provider';
import { IRcFile, IUploadFile, STATUS_NOT_YET } from './interface';

export const getUid = (): string => `${defaultRootPrefixCls}-upload-${Date.now()}`;

export const fileToObject = (file: IRcFile): IUploadFile => ({
  ...file,
  lastModified: file.lastModified,
  lastModifiedDate: file.lastModifiedDate,
  name: file.name,
  size: file.size,
  type: file.type,
  uid: file.uid,
  percent: 0,
  originFile: file,
});

export const getEmptyFileObj = (): IUploadFile => ({
  uid: getUid(),
  size: 0,
  name: '本地上传',
  type: '$empty-file',
  status: STATUS_NOT_YET,
});

export const requestImage = (src: string, cors?: 'anonymous' | 'use-credentials'): Promise<HTMLImageElement> =>
  new Promise((resolve, reject) => {
    const img = new Image();

    img.addEventListener('load', () => resolve(img));

    img.addEventListener('error', (err) => reject(err));

    if (cors) {
      img.setAttribute('crossOrigin', cors);
    }
    img.src = src;
  });
/**
 *
 * @param {File | Blob} file 图片文件
 * @returns {Promise<string> } 返回值是图片的 base64 字符串
 */
export const imageFile2DataUrl = (file: File | Blob): Promise<string> =>
  new Promise((resolve, reject) => {
    const canvas = document.createElement('canvas');
    const url = window.URL.createObjectURL(file);
    requestImage(url)
      .then((img) => {
        const { width, height } = img;
        canvas.width = width;
        canvas.height = height;

        const ctx = canvas.getContext('2d');
        ctx?.drawImage(img, 0, 0, width, height);
        const dataUrl = canvas.toDataURL();
        resolve(dataUrl);
      })
      .catch(reject);
  });

/**
 * 图片 base64 转为 File 格式
 */
export const dataUrl2ImageFile = (dataUrl = ''): File => {
  const arr = dataUrl.split(',');
  const mime = arr[0].match(/:(.*?);/)?.[1];
  const bstr = atob(arr[1] ?? '');
  let n = bstr.length;
  const u8arr = new Uint8Array(n);
  while (n) {
    n -= 1;
    u8arr[n] = bstr.charCodeAt(n);
  }
  return new File([u8arr], 'web-image', { type: mime });
};

/**
 *  网络 url 图片获取
 * @param {string} url
 */
export const fetchImageFileFromUrl = (
  url: string
): Promise<{
  originFile: IRcFile;
  dataUrl: string;
}> =>
  new Promise((resolve, reject) => {
    const canvas = document.createElement('canvas');
    requestImage(url, 'anonymous')
      .then((img) => {
        const { width, height } = img;
        canvas.width = width;
        canvas.height = height;

        const ctx = canvas.getContext('2d');
        ctx?.drawImage(img, 0, 0, width, height);
        let dataUrl = '';
        try {
          dataUrl = canvas.toDataURL();
        } catch (error) {
          reject(error);
        }
        const file = dataUrl2ImageFile(dataUrl);
        (file as IRcFile).uid = getUid();
        resolve({
          originFile: file as IRcFile,
          dataUrl,
        });
      })
      .catch((err) => reject(err));
  });

export const isOnlyAcceptImg = (accept?: string): boolean =>
  accept
    ? accept
        .split(',')
        .map((_) => _.trim())
        .filter((_) => !/((image\/)|\.)(dwg|dxf|gif|jp2|jpeg|jpg|jpe|png|svf|tif|tiff)|image\/\*/.test(_)).length === 0
    : false;

export const isImageFile = (file: IUploadFile): boolean => file.type.startsWith('image/');

/**
 * 暂时没有用起来，文件头在处理 docx/xlsx 等微软文件时候的区分方法暂时没写
 * @param file
 */
export const getFileType = (file: File | Blob): Promise<string | undefined> =>
  new Promise((resolve) => {
    const fr = new FileReader();

    fr.addEventListener('load', (e: any) => {
      const view = new DataView(e.target.result);
      const first4Byte = view.getUint32(0, false);
      const hexValue = Number(first4Byte).toString(16).toUpperCase();

      switch (hexValue) {
        case 'FFD8FFE0':
        case 'FFD8FFE1':
        case 'FFD8FFE2':
        case 'FFD8FFE3':
          resolve('JPEG/JPG');
          break;
        default:
          resolve(undefined);
          break;
      }
    });

    fr.readAsArrayBuffer(file);
  });
