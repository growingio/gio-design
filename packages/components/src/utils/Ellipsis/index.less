.gio-ui-ellipsis {
  display: inline-block;
  display: -webkit-box;
  overflow: hidden;
  text-overflow: ellipsis;
  word-break: break-all;
  -webkit-line-clamp: 2;
  &__line--2 {
    -webkit-line-clamp: 2;
  }
  &__line--1 {
    -webkit-line-clamp: 1;
  }
}
.gio-ui-single-line-ellipsis {
  display: inline-block;
  width: 100%;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  word-wrap: normal;
}
