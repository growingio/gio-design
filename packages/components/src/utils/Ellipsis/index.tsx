import * as React from 'react';
import { Tooltip } from 'antd';

interface P {
  line?: number;
  title?: React.ReactNode;
  maxWidth?: string;
  style?: object;
  placement?:
    | 'top'
    | 'left'
    | 'right'
    | 'bottom'
    | 'topLeft'
    | 'topRight'
    | 'bottomLeft'
    | 'bottomRight'
    | 'leftTop'
    | 'leftBottom'
    | 'rightTop'
    | 'rightBottom';
}

class Ellipsis extends React.PureComponent<P> {
  public state = {
    tooltip: false,
  };
  private textRef: HTMLSpanElement | null = null;
  public componentDidMount() {
    this.detectTooltip();
  }
  public componentDidUpdate(prevProps: P) {
    if (prevProps.title !== this.props.title) {
      this.detectTooltip();
    }
  }
  public render() {
    const { children, line, title, maxWidth, style, placement } = this.props;
    const _style = { maxWidth, ...style };
    if (!line) {
      return (
        <Tooltip
          title={this.state.tooltip ? title : ''}
          placement={placement}
          overlayStyle={{ wordWrap: 'break-word' }}
        >
          <span className='gio-ui-single-line-ellipsis' ref={this.setTextRef} style={_style}>
            {children}
          </span>
        </Tooltip>
      );
    }
    return (
      <Tooltip title=''>
        <span
          className={`gio-ui-ellipsis gio-ui-ellipsis__line--${line}`}
          style={{
            boxOrient: 'vertical',
            lineClamp: line,
            ..._style,
          }}
          ref={this.setTextRef}
        >
          {children}
        </span>
      </Tooltip>
    );
  }
  private setTextRef = (ref: HTMLSpanElement) => {
    if (ref) {
      this.textRef = ref;
    } else {
      this.textRef = null;
    }
  };
  private detectTooltip = () => {
    this.setTooltip();
    // IE、Firefox下多行省略不兼容，限制行数。
    this.limitLineNumbers();
  };
  private setTooltip = () => {
    let tooltip = false;
    if (this.textRef && this.textRef.scrollWidth > this.textRef.clientWidth) {
      tooltip = true;
    }
    if (tooltip !== this.state.tooltip) {
      this.setState({ tooltip });
    }
  };
  private limitLineNumbers = () => {
    const { line } = this.props;
    if (!line) {
      return;
    }
    if (this.textRef) {
      const lineHeight = parseInt(getComputedStyle(this.textRef)['line-height' as any], 10);
      if (typeof lineHeight === 'number' && !isNaN(lineHeight)) {
        this.textRef.style.maxHeight = line * lineHeight + 'px';
      }
    }
  };
}

export default Ellipsis;
