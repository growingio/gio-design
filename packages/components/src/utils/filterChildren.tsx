import React from 'react';

export type IFilterChildren = (
  children: React.ReactNode,
  filter: (child: React.ReactNode) => boolean
) => React.ReactNodeArray;

const filterChildren: IFilterChildren = (children, filter) => {
  const result: React.ReactNodeArray = [];

  React.Children.forEach(children, (child) => {
    if (filter(child)) {
      result.push(child);
    }
  });

  return result;
};

export default filterChildren;
