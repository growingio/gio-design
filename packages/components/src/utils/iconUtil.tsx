import * as React from 'react';
import CaretDownOutlined from '@gio-design/icons/es/CaretDownOutlined';
import LoadingOutlined from '@gio-design/icons/es/LoadingOutlined';
import CheckOutlined from '@gio-design/icons/es/CheckOutlined';
import Close from '@gio-design/icons/es/Close';
import CloseCircleFilled from '@gio-design/icons/es/CloseCircleFilled';
import SearchOutlined from '@gio-design/icons/es/SearchOutlined';

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export default function getIcons({
  suffixIcon,
  clearIcon,
  menuItemSelectedIcon,
  removeIcon,
  loading,
  multiple,
  prefixCls,
}: {
  suffixIcon?: React.ReactNode;
  clearIcon?: React.ReactNode;
  menuItemSelectedIcon?: React.ReactNode;
  removeIcon?: React.ReactNode;
  loading?: boolean;
  multiple?: boolean;
  prefixCls: string;
}) {
  // Clear Icon
  let mergedClearIcon = clearIcon;
  if (!clearIcon) {
    mergedClearIcon = <CloseCircleFilled size='12px' />;
  }

  // Arrow item icon
  let mergedSuffixIcon = null;
  if (suffixIcon !== undefined) {
    mergedSuffixIcon = suffixIcon;
  } else if (loading) {
    mergedSuffixIcon = <LoadingOutlined rotating size='12px' />;
  } else {
    const iconCls = `${prefixCls}-suffix`;
    mergedSuffixIcon = ({ open, showSearch }: { open: boolean; showSearch: boolean }) => {
      if (open && showSearch) {
        return <SearchOutlined className={iconCls} size='12px' />;
      }
      return <CaretDownOutlined className={iconCls} size='12px' />;
    };
  }

  // Checked item icon
  let mergedItemIcon = null;
  if (menuItemSelectedIcon !== undefined) {
    mergedItemIcon = menuItemSelectedIcon;
  } else if (multiple) {
    mergedItemIcon = <CheckOutlined size='12px' />;
  } else {
    mergedItemIcon = null;
  }

  let mergedRemoveIcon = null;
  if (removeIcon !== undefined) {
    mergedRemoveIcon = removeIcon;
  } else {
    mergedRemoveIcon = <Close size='12px' />;
  }

  return {
    clearIcon: mergedClearIcon,
    suffixIcon: mergedSuffixIcon,
    itemIcon: mergedItemIcon,
    removeIcon: mergedRemoveIcon,
  };
}
