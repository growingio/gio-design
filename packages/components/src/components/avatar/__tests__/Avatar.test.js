import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Avatar from '../Avatar';
import Dropdown from '../../dropdown';
import '../../../../es/components/avatar/style/index.css';
import { waitForComponentToPaint } from '../../../utils/test';
import image from './icon.jpeg';

describe('Testing Avatar', () => {
  it('should be stable', () => {
    const wrapper = renderer.create(<Avatar>快照</Avatar>).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it('should be mount, setProps, unmount with no error', () => {
    expect(() => {
      const wrapper = mount(<Avatar />);
      wrapper.setProps({
        size: 'small',
        src: '123',
        omit: false,
      });
      wrapper.setProps({
        size: 'large',
        omit: true,
      });
      wrapper.unmount();
    }).not.toThrow();
  });

  test('props src', () => {
    const wrapper = mount(<Avatar src={image}>li</Avatar>).children();
    expect(wrapper.childAt(0).type()).toBe('img');
    expect(wrapper.childAt(0).text()).not.toBe('l');
    const noSrcWrapper = mount(<Avatar size="small">li</Avatar>).children();
    expect(noSrcWrapper.childAt(0).type()).toBe('span');
    expect(noSrcWrapper.childAt(0).text()).toBe('l');
  });

  test('props size', () => {
    expect(mount(<Avatar size="small" />).exists('.gio-avatar-sm')).toBe(true);
    expect(mount(<Avatar />).exists('.gio-avatar-df')).toBe(true);
    expect(mount(<Avatar size="large" />).exists('.gio-avatar-lg')).toBe(true);
    expect(mount(<Avatar size="huge" />).exists('.gio-avatar-hg')).toBe(true);
  });

  test('prop droppable', () => {
    expect(mount(<Avatar size="small" droppable />).exists('.gio-avatar-droppable')).toBe(true);
  });

  test('props omit', () => {
    const wrapper = mount(<Avatar>这是一个很长的文字</Avatar>);
    expect(wrapper.childAt(0).text()).toBe('这');
    wrapper.setProps({ omit: false });
    expect(wrapper.childAt(0).text()).toBe('这是一个很长的文字');
  });

  test('props displayTooltip', () => {
    const wrapper = mount(<Avatar displayTooltip>这是一个很长的文字</Avatar>);
    expect(wrapper.find('.gio-avatar').at(0).text()).toBe('这');
    wrapper.find('.gio-avatar').at(0).simulate('mouseenter');
    expect(wrapper.find('.gio-tooltip-inner-title').text()).toBe('这是一个很长的文字');
  });

  test('props placement', async () => {
    const wrapper = mount(<Avatar displayTooltip>这是一个很长的文字</Avatar>);
    wrapper.setProps({ placement: 'top' });
    wrapper.find('.gio-avatar').at(0).simulate('mouseenter');
    await waitForComponentToPaint(wrapper);
    expect(wrapper.exists('.gio-tooltip-placement-top')).toBe(true);
  });

  it('can accept dropdown trigger Mouse Event', () => {
    const wrapper = mount(
      <Dropdown overlay={<div>11</div>}>
        <Avatar>li</Avatar>
      </Dropdown>
    );
    wrapper.find('.gio-avatar').at(0).simulate('click');
    expect(wrapper.exists('.gio-dropdown')).toBe(true);
  });
});
