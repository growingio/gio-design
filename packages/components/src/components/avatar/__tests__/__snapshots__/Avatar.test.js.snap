// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Avatar should be stable 1`] = `
<span
  className="gio-avatar gio-avatar-df"
>
  <span
    style={
      Object {
        "transform": "scale(1)",
      }
    }
  >
    å¿«
  </span>
</span>
`;
